source/d1amp.cpp

```diff
@@ -57,6 +53,37 @@ bool D1Amp::load(QString filePath, int allocate)
     return true;
 }
 
+bool D1Amp::save(SaveAsParam *params)
+{
+    QString filePath = this->getFilePath();
+    if (params != nullptr && !params->ampFilePath.isEmpty()) {
+        filePath = params->ampFilePath;
+        if (QFile::exists(filePath)) {
+            QMessageBox::StandardButton reply;
+            reply = QMessageBox::question(nullptr, "Confirmation", "Are you sure you want to overwrite the AMP file?", QMessageBox::Yes | QMessageBox::No);
+            if (reply != QMessageBox::Yes) {
+                return false;
+            }
+        }
+    }
+
+    QFile outFile = QFile(filePath);
+    if (!outFile.open(QIODevice::WriteOnly | QFile::Truncate)) {
+        QMessageBox::critical(nullptr, "Error", "Failed open file: " + filePath);
+        return false;
+    }
```

Comment:
Looks like there are 7 almost identical copies of this. Could we maybe extract it in to a common function if we change `"Are you sure you want to overwrite the AMP file?"` to something like `"Are you sure you want to overwrite {}?", filePath´?

Reply:
I don't see much point in it. It does not really cover a real functionality, it just bunch of lines which are common. Besides `filePath`, `params` and `params->ampFilePath` should be passed to that common function. At that point the whole function is gutted.

Reply:
To avoid duplication, make the save functions more concise and specific to the file data.
It looks like the function could return `std::optional<QFile>`, in case of an empty response the save would do an early return.

Something roughly along this line:
```cpp
std::optional<QFile> getValidSaveOutput(QString filePath, QString selectedPath) {
    if (!selectedPath.isEmpty()) {
        filePath = selectedPath;
        if (QFile::exists(filePath)) {
            QMessageBox::StandardButton reply;
            reply = QMessageBox::question(nullptr, "Confirmation", "Are you sure you want to overwrite " + filePath + "?", QMessageBox::Yes | QMessageBox::No);
            if (reply != QMessageBox::Yes) {
                return {};
            }
        }
    }

    QFile outFile = QFile(filePath);
    if (!outFile.open(QIODevice::WriteOnly | QFile::Truncate)) {
        QMessageBox::critical(nullptr, "Error", "Failed open file: " + filePath);
        return {};
    }
    
    return outFile;
}
```

And then the call becomes:
```cpp
    QString selectedPath = params != nullptr ? params->ampFilePath : "";
    std::optional<QFile> outFile = getValidSaveOutput(this->getFilePath(), selectedPath);
    if (!outFile) {
        return false;
    }
```

Reply:
The current build error can probably be fixed by `QDataStream out(&outFile);` -> `QDataStream out(&outFile->get());` (haven't tested it, so  mostly a quess)